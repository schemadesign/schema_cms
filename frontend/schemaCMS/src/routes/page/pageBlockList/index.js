export { default as PageBlockList } from './pageBlockList.container';
