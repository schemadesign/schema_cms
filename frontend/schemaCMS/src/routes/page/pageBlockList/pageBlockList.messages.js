/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

export default defineMessages({
  blocks: {
    id: 'page.blockList.blocks',
    defaultMessage: '{length} {length, plural,one {Block} other {Blocks}}',
  },
  title: {
    id: 'page.blockList.title',
    defaultMessage: 'Blocks',
  },
  subTitle: {
    id: 'page.blockList.subTitle',
    defaultMessage: 'Overview',
  },
  save: {
    id: 'page.blockList.save',
    defaultMessage: 'Save',
  },
});
