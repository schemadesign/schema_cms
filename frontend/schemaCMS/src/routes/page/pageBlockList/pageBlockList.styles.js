import styled from 'styled-components';
import { Link as RouterLink } from 'react-router-dom';

export const Container = styled.div``;

export const CreateButtonContainer = styled.div`
  width: 20%;
`;

export const BlockCounter = styled.div`
  width: 60%;
  text-align: center;
  color: ${({ theme: { secondaryText } }) => secondaryText};
`;

export const Header = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-bottom: 24px;
`;

export const Empty = styled.div`
  width: 20%;
`;

export const Link = styled(RouterLink)`
  color: inherit;
  text-decoration: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

export const checkboxStyles = {
  width: '100%',
};

export const CheckboxContent = styled.div`
  display: flex;
  align-items: center;
`;

export const IconWrapper = styled.div`
  display: flex;
  cursor: move;
  touch-action: none;
`;

export const menuIconStyles = { width: 10, marginRight: 20 };
