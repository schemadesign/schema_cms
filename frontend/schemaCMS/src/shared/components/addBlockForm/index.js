export { AddBlockForm } from './addBlockForm.component';
