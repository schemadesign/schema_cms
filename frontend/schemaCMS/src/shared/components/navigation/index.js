export { NavigationContainer, BackArrowButton, PlusButton, BackButton, NextButton } from './navigation.component';
