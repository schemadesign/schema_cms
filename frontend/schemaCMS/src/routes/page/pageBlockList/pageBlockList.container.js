import { connect } from 'react-redux';
import { bindPromiseCreators, promisifyRoutine } from 'redux-saga-routines';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';
import { hot } from 'react-hot-loader';
import { compose } from 'ramda';
import { injectIntl } from 'react-intl';
import { withFormik } from 'formik';

import { PageBlockList } from './pageBlockList.component';
import { PageBlockRoutines, selectPageBlocks, selectTemporaryPageBlocks } from '../../../modules/pageBlock';
import { PageRoutines, selectPage } from '../../../modules/page';
import { selectUserRole } from '../../../modules/userProfile';
import { errorMessageParser, getMatchParam } from '../../../shared/utils/helpers';
import messages from '../../pageBlock/pageBlock.messages';

const mapStateToProps = createStructuredSelector({
  userRole: selectUserRole,
  temporaryPageBlocks: selectTemporaryPageBlocks,
  pageBlocks: selectPageBlocks,
  page: selectPage,
});

export const mapDispatchToProps = dispatch => ({
  ...bindPromiseCreators(
    {
      fetchPageBlocks: promisifyRoutine(PageBlockRoutines.fetchList),
      saveTemporaryBlocks: promisifyRoutine(PageBlockRoutines.saveTemporaryBlocks),
      setPageBlocks: promisifyRoutine(PageBlockRoutines.setBlocks),
      fetchPage: promisifyRoutine(PageRoutines.fetchOne),
    },
    dispatch
  ),
});

export default compose(
  hot(module),
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  injectIntl,
  withRouter,
  withFormik({
    enableReinitialize: true,
    mapPropsToValues: ({ pageBlocks }) => pageBlocks,
    handleSubmit: async (values, { props, setSubmitting, setErrors }) => {
      try {
        const pageId = getMatchParam(props, 'pageId');
        const blocks = values.map(({ id, isActive, name }, index) => ({
          isActive,
          name,
          id,
          execOrder: index,
        }));

        await props.setPageBlocks({ pageId, blocks });
      } catch (errors) {
        const { formatMessage } = props.intl;
        const errorMessages = errorMessageParser({ errors, messages, formatMessage });

        setErrors(errorMessages);
      } finally {
        setSubmitting(false);
      }
    },
  })
)(PageBlockList);
