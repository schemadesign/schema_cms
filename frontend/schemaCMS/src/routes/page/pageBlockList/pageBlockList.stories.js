import React from 'react';
import { storiesOf } from '@storybook/react';

import { PageBlockList } from './pageBlockList.component';
import { withTheme } from '../../../.storybook/decorators';
import { history, intl } from '../../../.storybook/helpers';
import { ROLES } from '../../../modules/userProfile/userProfile.constants';

export const defaultProps = {
  userRole: ROLES.ADMIN,
  values: [
    {
      name: 'block 1',
      isActive: true,
      id: 1,
    },
    {
      name: 'block 2',
      isActive: false,
      id: 2,
    },
  ],
  page: {
    folder: {
      id: '1',
    },
  },
  temporaryPageBlocks: [],
  saveTemporaryBlocks: Function.prototype,
  fetchPageBlocks: Function.prototype,
  fetchPage: Function.prototype,
  handleSubmit: Function.prototype,
  setValues: Function.prototype,
  isSubmitting: false,
  dirty: false,
  match: {
    params: {
      pageId: '1',
    },
  },
  intl,
  history,
};

const emptyBlocksProps = {
  ...defaultProps,
  pageBlocks: [],
};

storiesOf('Page|PageBlockList', module)
  .addDecorator(withTheme())
  .add('No data', () => <PageBlockList {...emptyBlocksProps} />)
  .add('Default', () => <PageBlockList {...defaultProps} />);
