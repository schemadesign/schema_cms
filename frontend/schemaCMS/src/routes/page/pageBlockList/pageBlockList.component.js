import React, { Fragment, PureComponent } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import Helmet from 'react-helmet';
import { always, append, equals, filter, ifElse, map, path, pipe, prop, propEq, reject, toString } from 'ramda';
import { Formik } from 'formik';
import { Form } from 'schemaUI';

import { BlockCounter, Container, CreateButtonContainer, Empty, Header, Link } from './pageBlockList.styles';
import messages from './pageBlockList.messages';
import { BackArrowButton, NavigationContainer, NextButton, PlusButton } from '../../../shared/components/navigation';
import { TopHeader } from '../../../shared/components/topHeader';
import { ContextHeader } from '../../../shared/components/contextHeader';
import { LoadingWrapper } from '../../../shared/components/loadingWrapper';
import { renderWhenTrue } from '../../../shared/utils/rendering';

const { CheckboxGroup, Checkbox } = Form;

export class PageBlockList extends PureComponent {
  static propTypes = {
    pageBlocks: PropTypes.array.isRequired,
    page: PropTypes.object.isRequired,
    fetchPageBlocks: PropTypes.func.isRequired,
    fetchPage: PropTypes.func.isRequired,
    setPageBlocks: PropTypes.func.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        pageId: PropTypes.string.isRequired,
      }),
    }),
    intl: PropTypes.shape({
      formatMessage: PropTypes.func.isRequired,
    }),
    history: PropTypes.shape({
      push: PropTypes.func.isRequired,
    }),
  };

  state = {
    loading: true,
    error: null,
  };

  async componentDidMount() {
    try {
      const pageId = this.getPageId();
      await this.props.fetchPage({ pageId });
      await this.props.fetchPageBlocks({ pageId });
      this.setState({ loading: false });
    } catch (error) {
      this.setState({ loading: false, error });
    }
  }

  getPageId = () => path(['match', 'params', 'pageId'], this.props);
  getDirectoryId = () => path(['page', 'directory', 'id'], this.props);

  handleCreateBlock = () => this.props.history.push(`/page/${this.getPageId()}/block/create`);
  handleShowPages = () => this.props.history.push(`/directory/${this.getDirectoryId()}`);

  handleSubmit = ({ pageBlocks }) => {
    const pageId = this.getPageId();
    const inactive = this.props.pageBlocks
      .filter(({ id }) => !pageBlocks.includes(id.toString()))
      .map(({ id }) => id.toString());

    this.props.setPageBlocks({ pageId, active: pageBlocks, inactive });
  };

  handleChange = ({ e, setFieldValue, pageBlocks }) => {
    const { value, checked } = e.target;
    const setPageBlocks = ifElse(
      equals(true),
      always(append(value, pageBlocks)),
      always(reject(equals(value), pageBlocks))
    );

    setFieldValue('pageBlocks', setPageBlocks(checked));
  };

  renderCheckbox = ({ id, name }, index) => (
    <Checkbox id={`checkbox-${index}`} value={id.toString()} key={index} isEdit>
      <Link to={`/block/${id}`}>{name}</Link>
    </Checkbox>
  );

  renderContent = () => {
    const { pageBlocks: list } = this.props;
    const { loading, error } = this.state;

    const activeBlocks = pipe(
      filter(propEq('isActive', true)),
      map(prop('id')),
      map(toString)
    )(list);

    return (
      <Formik initialValues={{ pageBlocks: activeBlocks }} onSubmit={this.handleSubmit}>
        {({ values: { pageBlocks }, setFieldValue, submitForm, dirty }) => {
          return (
            <Fragment>
              <LoadingWrapper loading={loading} error={error} noData={!list.length}>
                <CheckboxGroup
                  onChange={e => this.handleChange({ e, setFieldValue, pageBlocks })}
                  value={pageBlocks}
                  name="pageBlocks"
                  id="blocksCheckboxGroup"
                >
                  {list.map(this.renderCheckbox)}
                </CheckboxGroup>
              </LoadingWrapper>
              <NavigationContainer>
                <BackArrowButton type="button" onClick={this.handleShowPages} />
                <NextButton type="submit" onClick={submitForm} disabled={!dirty}>
                  <FormattedMessage {...messages.save} />
                </NextButton>
              </NavigationContainer>
            </Fragment>
          );
        }}
      </Formik>
    );
  };

  renderBlockCounter = renderWhenTrue(
    always(
      <BlockCounter>
        <FormattedMessage values={{ length: this.props.pageBlocks.length }} {...messages.blocks} />
      </BlockCounter>
    )
  );

  render() {
    const { loading } = this.state;
    const headerTitle = <FormattedMessage {...messages.title} />;
    const headerSubtitle = <FormattedMessage {...messages.subTitle} />;

    return (
      <Container>
        <Helmet title={this.props.intl.formatMessage(messages.title)} />
        <TopHeader headerTitle={headerTitle} headerSubtitle={headerSubtitle} />
        <ContextHeader title={headerTitle} subtitle={headerSubtitle} />
        <Header>
          <CreateButtonContainer>
            <PlusButton onClick={this.handleCreateBlock} />
          </CreateButtonContainer>
          {this.renderBlockCounter(!loading)}
          <Empty />
        </Header>
        {this.renderContent()}
      </Container>
    );
  }
}
