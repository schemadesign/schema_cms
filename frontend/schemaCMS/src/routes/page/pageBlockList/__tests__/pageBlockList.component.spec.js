import React from 'react';
import { shallow } from 'enzyme';
import { Formik } from 'formik';

import { PageBlockList } from '../pageBlockList.component';
import { Link } from '../pageBlockList.styles';
import { defaultProps } from '../pageBlockList.stories';
import { BackArrowButton, PlusButton } from '../../../../shared/components/navigation';

describe('PageBlockList: Component', () => {
  const component = props => <PageBlockList {...defaultProps} {...props} />;

  const render = (props = {}) => shallow(component(props));

  it('should render correctly with loader', () => {
    const wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly', async () => {
    const props = {
      fetchPage: jest.fn().mockReturnValue(Promise.resolve()),
      fetchPageBlocks: jest.fn().mockReturnValue(Promise.resolve()),
    };
    const wrapper = render(props);
    await Promise.resolve();
    await Promise.resolve();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly form', async () => {
    const props = {
      fetchPage: jest.fn().mockReturnValue(Promise.resolve()),
      fetchPageBlocks: jest.fn().mockReturnValue(Promise.resolve()),
    };
    const wrapper = render(props);

    await Promise.resolve();
    await Promise.resolve();

    const form = wrapper.find(Formik).dive();

    expect(form).toMatchSnapshot();
  });

  it('should fetch list', async () => {
    jest.spyOn(defaultProps, 'fetchPage');
    jest.spyOn(defaultProps, 'fetchPageBlocks');
    render();
    await Promise.resolve();
    await Promise.resolve();
    expect(defaultProps.fetchPage).toBeCalledWith({ pageId: '1' });
    expect(defaultProps.fetchPageBlocks).toBeCalledWith({ pageId: '1' });
  });

  it('should go to page list page', async () => {
    const props = {
      fetchPage: jest.fn().mockReturnValue(Promise.resolve()),
      fetchPageBlocks: jest.fn().mockReturnValue(Promise.resolve()),
    };
    jest.spyOn(defaultProps.history, 'push');
    const wrapper = render(props);
    await Promise.resolve();
    await Promise.resolve();
    wrapper.find(BackArrowButton).simulate('click');

    expect(defaultProps.history.push).toBeCalledWith('/folder/1');
  });

  it('should go to create block', () => {
    jest.spyOn(defaultProps.history, 'push');
    const wrapper = render();
    wrapper.find(PlusButton).simulate('click');

    expect(defaultProps.history.push).toBeCalledWith('/page/1/block/create');
  });

  it('should have link to block', async () => {
    const props = {
      fetchPage: jest.fn().mockReturnValue(Promise.resolve()),
      fetchPageBlocks: jest.fn().mockReturnValue(Promise.resolve()),
    };
    const wrapper = render(props);
    await Promise.resolve();
    await Promise.resolve();
    const link = wrapper.find(Link).at(1);

    expect(link.props('to').to).toBe('/block/2');
  });
});
