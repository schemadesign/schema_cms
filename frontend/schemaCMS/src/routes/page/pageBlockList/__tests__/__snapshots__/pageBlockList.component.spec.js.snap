// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageBlockList: Component should render correctly 1`] = `
<styled.div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="Blocks"
  />
  <TopHeader
    headerSubtitle={
      <FormattedMessage
        defaultMessage="Overview"
        id="page.blockList.subTitle"
        values={Object {}}
      />
    }
    headerTitle={
      <FormattedMessage
        defaultMessage="Blocks"
        id="page.blockList.title"
        values={Object {}}
      />
    }
    primaryMenuItems={Array []}
    secondaryMenuItems={Array []}
  />
  <ContextHeader
    subtitle={
      <FormattedMessage
        defaultMessage="Overview"
        id="page.blockList.subTitle"
        values={Object {}}
      />
    }
    title={
      <FormattedMessage
        defaultMessage="Blocks"
        id="page.blockList.title"
        values={Object {}}
      />
    }
  />
  <styled.div>
    <styled.div>
      <PlusButton
        hideOnDesktop={false}
        onClick={[Function]}
      />
    </styled.div>
    <styled.div>
      <FormattedMessage
        defaultMessage="{count, plural,one {# Block} other {# Blocks}}"
        id="page.blockList.blocks"
        values={
          Object {
            "count": 2,
          }
        }
      />
    </styled.div>
    <styled.div />
  </styled.div>
  <Formik
    enableReinitialize={true}
    initialValues={
      Object {
        "blocks": Array [
          "1",
        ],
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <Component />
  </Formik>
</styled.div>
`;

exports[`PageBlockList: Component should render correctly form 1`] = `
<ContextProvider
  value={
    Object {
      "dirty": false,
      "errors": Object {},
      "handleBlur": [Function],
      "handleChange": [Function],
      "handleReset": [Function],
      "handleSubmit": [Function],
      "initialValues": Object {
        "blocks": Array [
          "1",
        ],
      },
      "isSubmitting": false,
      "isValid": false,
      "isValidating": false,
      "registerField": [Function],
      "resetForm": [Function],
      "setError": [Function],
      "setErrors": [Function],
      "setFieldError": [Function],
      "setFieldTouched": [Function],
      "setFieldValue": [Function],
      "setFormikState": [Function],
      "setStatus": [Function],
      "setSubmitting": [Function],
      "setTouched": [Function],
      "setValues": [Function],
      "status": undefined,
      "submitCount": 0,
      "submitForm": [Function],
      "touched": Object {},
      "unregisterField": [Function],
      "validate": undefined,
      "validateField": [Function],
      "validateForm": [Function],
      "validateOnBlur": true,
      "validateOnChange": true,
      "validationSchema": undefined,
      "values": Object {
        "blocks": Array [
          "1",
        ],
      },
    }
  }
>
  <LoadingWrapper
    error={null}
    loading={false}
    noData={false}
    noDataContent="0 Blocks"
  >
    <Component
      id="blocksCheckboxGroup"
      name="pageBlocks"
      onChange={[Function]}
      value={
        Array [
          "1",
        ]
      }
    >
      <Component
        id="checkbox-0"
        isEdit={true}
        key="0"
        value="1"
      >
        <Styled(Link)
          to="/block/1"
        >
          block 1
        </Styled(Link)>
      </Component>
      <Component
        id="checkbox-1"
        isEdit={true}
        key="1"
        value="2"
      >
        <Styled(Link)
          to="/block/2"
        >
          block 2
        </Styled(Link)>
      </Component>
    </Component>
  </LoadingWrapper>
  <NavigationContainer
    fixed={true}
    hideOnDesktop={false}
    right={false}
  >
    <BackArrowButton
      onClick={[Function]}
      type="button"
    />
    <NextButton
      disabled={true}
      loading={false}
      onClick={[Function]}
      type="submit"
    >
      <FormattedMessage
        defaultMessage="Save"
        id="page.blockList.save"
        values={Object {}}
      />
    </NextButton>
  </NavigationContainer>
</ContextProvider>
`;

exports[`PageBlockList: Component should render correctly with loader 1`] = `
<styled.div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="Blocks"
  />
  <TopHeader
    headerSubtitle={
      <FormattedMessage
        defaultMessage="Overview"
        id="page.blockList.subTitle"
        values={Object {}}
      />
    }
    headerTitle={
      <FormattedMessage
        defaultMessage="Blocks"
        id="page.blockList.title"
        values={Object {}}
      />
    }
    primaryMenuItems={Array []}
    secondaryMenuItems={Array []}
  />
  <ContextHeader
    subtitle={
      <FormattedMessage
        defaultMessage="Overview"
        id="page.blockList.subTitle"
        values={Object {}}
      />
    }
    title={
      <FormattedMessage
        defaultMessage="Blocks"
        id="page.blockList.title"
        values={Object {}}
      />
    }
  />
  <styled.div>
    <styled.div>
      <PlusButton
        hideOnDesktop={false}
        onClick={[Function]}
      />
    </styled.div>
    <styled.div />
  </styled.div>
  <Formik
    enableReinitialize={true}
    initialValues={
      Object {
        "blocks": Array [
          "1",
        ],
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <Component />
  </Formik>
</styled.div>
`;
